controllers/metal3.io/baremetalhost_controller_test.go

```diff
@@ -1854,3 +1861,66 @@ func TestUpdateRAID(t *testing.T) {
 		})
 	}
 }
+
+// ReconcileHelper is a test helper class to support verifying
+// specific reconciliation sequences step-by-step and for
+// improving test readability
+type ReconcileHelper struct {
+	t          *testing.T
+	host       *metal3v1alpha1.BareMetalHost
+	reconciler *BareMetalHostReconciler
+	request    reconcile.Request
+}
+
+func newReconcileHelper(t *testing.T, host *metal3v1alpha1.BareMetalHost) *ReconcileHelper {
+	return &ReconcileHelper{
+		t:          t,
+		host:       host,
+		reconciler: newTestReconciler(host),
+		request:    newRequest(host),
+	}
+}
+
+func (r *ReconcileHelper) verifyNextReconcile(expectedState metal3v1alpha1.ProvisioningState, expectedRequeue bool, expectedRequeueAfter int) {
+	res, err, host := tryReconcileSingleStep(r.reconciler, r.host, r.request)
+	assert.NoError(r.t, err)
+	assert.Equal(r.t, expectedRequeue, res.Requeue)
+	assert.Equal(r.t, time.Duration(expectedRequeueAfter), res.RequeueAfter)
+	assert.Equal(r.t, expectedState, host.Status.Provisioning.State)
+}
+
+func (r *ReconcileHelper) verifyNextReconcileOk(expectedState metal3v1alpha1.ProvisioningState) {
+	r.verifyNextReconcile(expectedState, true, 0)
+}
+
+func (r *ReconcileHelper) verifyNextReconcileStops(expectedState metal3v1alpha1.ProvisioningState) {
+	r.verifyNextReconcile(expectedState, false, 0)
+}
+
+func (r *ReconcileHelper) verifyNextReconcileError(expectedState metal3v1alpha1.ProvisioningState, expectedErrorMessage string) {
+	_, err, host := tryReconcileSingleStep(r.reconciler, r.host, r.request)
+	assert.EqualError(r.t, err, expectedErrorMessage)
+	assert.Equal(r.t, expectedState, host.Status.Provisioning.State)
+}
+
+func (r *ReconcileHelper) doDeleteHost() {
+	now := metav1.Now()
+	r.host.DeletionTimestamp = &now
+	r.host.Finalizers = []string{}
```

Comment:
Shouldn't we be testing that the controller can remove the finalizer by itself? The actual bug is that it wasn't doing that so it would be good to test it, although in practice we know that if we can get the the Deleting state it should be able to.

Reply:
Thanks that's a good point, let me recheck it, as the DeletionTimestamp should be sufficient